src/libcollections/str.rs

```diff
@@ -1506,6 +1511,28 @@ pub trait StrExt: Index<RangeFull, Output = str> {
     fn trim_right(&self) -> &str {
         UnicodeStr::trim_right(&self[..])
     }
+
+    /// Returns the lowercase equivalent of this string.
+    ///
+    /// # Examples
+    ///
+    /// let s = "HELLO";
+    /// assert_eq!(s.to_lowercase(), "hello");
+    #[unstable(feature = "collections")]
+    fn to_lowercase(&self) -> String {
+        self[..].chars().flat_map(|c| c.to_lowercase()).collect()
```

Comment:
`FlatMap::size_hint` can make no assumption about the underlying iterators, so it has to be pessimistic. `.collect()` here could be replaced with `let mut s = String::with_capacity(self.len()); s.extend(…); s`. (Same in `to_uppercase`.)